### ISSUE TITLE
v8_r6_Josh7: Mobs spawning in OTG worlds being replaced by mobs from other mods sharing the same name

### ISSUE THREAD

**Issue by Muugay** [PG85/OpenTerrainGenerator#339]
#### OTG Plugin or Mod

- [ X] OTG Mod for Forge
- [ ] OTG Plugin for Spigot

#### OTG Version

- [ ] 1.12.2_v8_r1
- [ ] 1.12.2_v7_DEV_r26
- [ ] 1.12.2-v6
- [ ] 1.12.2-v5
- [ ] 1.11.2-v8
- [ ] 1.11.2-v9
- [ ] 1.10.2-v21
- [X ] Other - OpenTerrainGenerator-1.12.2_-_v8_r6_Josh7

#### Forge or Spigot Versions

Forge version 14.23.5.2838

#### What other plugins are you running?

Forge Mod Loader             
 Advanced Hook Launchers      
 Aether Legacy                
 Aether Continuation          
 AI Improvements              
 Akashic Tome                 
 Antique Atlas                
 Antique Atlas Overlay        
 Atlas Extras                 
 AutoRegLib                   
 Baubles                      
 Better Advancements          
 Cyber's Better Animal Models 
 Better Animals Plus          
 YUNG's Better Caves          
 Better Questing              
 BiblioCraft                  
 Bigger Packets, please !     
 Biome Bundle                 
 Bloodmoon                    
 Bountiful                    
 Bountiful Baubles            
 Standard Expansion           
 Building Gadgets             
 Chameleon                    
 Charm                        
 Chisel                       
 Chunk Animator               
 Clumps                       
 Colossal Chests              
 Controlling                  
 Cooking for Blockheads       
 Coralreef                    
 CraftTweaker2                
 CraftTweaker JEI Support     
 CT-GUI                       
 CTM                          
 FlameLib                     
 Customized Dungeon Loot      
 Cyclic                       
 Cyclops Core                 
 Dimensional Doors            
 Doggy Talents                
 §3Dynamic Surroundings       
 DynamicSurroundingsCore      
 DungeonTactics               
 Electroblob's Wizardry       
 Embers Rekindled             
 Enchantment Descriptions     
 Engineer's Workshop Reborn   
 Exotic Birds                 
 Meldexun's Extra Spells      
 Extra Alchemy                
 Extra Utilities 2            
 Eyes in the Darkness         
 FastWorkbench                
 FastFurnace                  
 First Aid                    
 FoamFix                      
 FoamFixCore                  
 Minecraft Forge              
 Forge Endertech              
 Shadowfacts' Forgelin        
 FTB Utilities: Backups       
 FTB Guides                   
 FTB Library                  
 FTB Money                    
 FTB Quests                   
 FTB Utilities                
 Gotta Go Fast                
 Ice and Fire                 
 InControl                    
 Infernal Mobs                
 InGame Info XML              
 Inventory Pets               
 Inventory Tweaks             
 Item Filters                 
 IvToolkit                    
 Just Enough Items            
 Just Enough Resources        
 LibraryEx                    
 Living Enchantment           
 LLibrary                     
 Login Shield                 
 LootTweaker                  
 Lost Aether Content          
 LunatriusCore                
 Lycanites Mobs               
 McJtyLib                     
 Minecraft Coder Pack         
 Mega Loot                    
 Minecraft                    
 Minerva Library              
 Mo' Bends                    
 Morpheus                     
 Mouse Tweaks                 
 Mowzie's Mobs                
 Mystical Mechanics           
 Neat                         
 NetherEx                     
 Notes                        
 OpenBlocks                   
 OpenModsLib                  
 OpenModsLib Core             
 Open Terrain Generator       
 Ordinary Coins               
 OreExcavation                
 §3OreLib Support Library     
 OTG Core                     
 Patchouli                    
 Phosphor Lighting Engine     
 PiTweaks                     
 Placeable Items Mod          
 Potion Core                  
 Progressive Bosses           
 Jack-O'-Launcher             
 Project: Vibrant Journeys    
 Quality Tools                
 Quark                        
 Quark Oddities               
 Random Things                
 ReAuth                       
 Reborn Core                  
 Recurrent Complex            
 Reskillable                  
 Roguelike Dungeons           
 Rope Bridge                  
 Rough Tweaks                 
 Nomadic Tents                
 Rustic                       
 Scaling Health               
 Silent Lib                   
 Sound Filters                
 Spartan Shields              
 SpawnerControl               
 Spawn Temple                 
 Stackie                      
 Stacksize Mod                
 SwingThroughGrass            
 Storage Drawers              
 Straw Golem                  
 Tell Me                      
 TerraCart Reloaded           
 The Betweenlands             
 Corail Tombstone             
 Totemic                      
 Tumbleweed                   
 The Twilight Forest          
 VanillaFix                   
 V-Tweaks                     
 Waila                        
 Waystones                    
 Wearable Backpacks           
 World Primer                 
 Trinkets and Baubles         
 Reliquary                    

---This is the list of mods in which the issue is not present, please read the how to reproduce section---
#### Description

When the issue is present, in OTG worlds mod mobs can spawn as other mod's mobs if that other mod's mob shares the same or similar name. The biggest example of this is with mowzies mobs naga being replaced by twilight forest naga but other instances have been present such as betteranimals squirrels being replaced by twilight forest squirrels (in previous versions of my modpack, current version I disabled betteranimals squirrels and manually added TF squirrels to my overworld so that it could be consistent via the "incontrol mod).

After extensive testing to resolve this issue, I've come to the conclusion that the issue is only present if a certain amount of entity entries is present during the minecraft loading phase; and only in OTG worlds. I initially thought it might've been based on the amount of mods loaded, or based on if it's a world genned through non vanilla means; but after testing via adding content adding mods that don't add entities, and testing via substituting OTG/biome bundle with another mod such as RTG  I can confirm neither of these are the cause.

As the amount of entities available when minecraft launches appears to be what triggers the issue, here are the steps to reproduce:

1.Put attached folders into minecraft directory
2.Launch minecraft
3.Create world with cheats enabled
4. /time set night 
5./otg tp Stone Beach
6./otg biome -m to confirm that under "Monster:" it is spawning "EntityNaga" (this is the mowzies variant)
7.Close minecraft
8.Add any entity adding mod to the mod's folder (in "Mods to add (Either or Both)" there are 2 different mods that can work for this)
9. Repeat steps 2>5
10./otg biome -m to confirm that under "Monster:" it is spawning "EntityTFNaga" instead
11.(Optional) Delete any of the added entity adding mods and repeat steps 2>7 to confirm that the issue is no longer present

I've attached the mod/configs/script folders from my minecraft instance in which the issue does not exist and included a folder labeled "Mods to add (Either or Both)" that has some example mods that you can add to the mod load order to trigger this issue, but I can assure you it'd occur with any other mods you can think of that add entities. 

In conclusion:
Neither of the mods included in "Mods to add (Either or Both)", and none of the mods I have currently installed with the exception of incontrol change minecrafts spawning mechanics. This combined with the fact that the issue is 100% only present in OTG worlds heavily implies the issue is on OTGs end. I've been working on resolving this issue for 3 days now and honestly as I am not heavily knowledgeable with coding this is the only conclusion I could come up with.

Please let me know if there's anything else I can do to help in order to resolve this issue.

Attached folders:
https://www.dropbox.com/s/4z1qkydbkh1xqur/OTG%20Naga%20Issue.rar?dl=0

**Comment by PG85** [comment]
Hey @Muugay , thanks for reporting this. Sounds like this may be related to OTG's mob inheritance mechanics, which may not be able to differentiate between mods with the same name. Moving this to backlog for now, will assign a proper milestone as soon as v8 is done.

**Comment by MCPitman** [comment]
I actually tried to help with this in the discord and I could not replicate it at all - looks like the guy got to the bottom of the issue anyway - quote from the discord channel:

```
Muugi:
It's mocreatures, apparently just the basemod has some sort of code that either looks at drzharks customspawner configs or has similar code to it
I disabled custommobspawners as I recalled in V6 it warned you about possible issues, but it seems that even with it disabled just mocreatures might affect non mocreature spawning
Since looking at the different configs, stuff like fox from betteranimals says Spawn: biomegropus <Fox_Default> rather than Spawn:biomegroups <betteranimals_fox:default>
```

**Comment by Gunner76th** [comment]
Is there any update on this, and if a fix is being put in place for V9  so that this no longer happens? In Roguelike Adventures and Dungeons we are starting to have TLF Naga spawn in the overworld because OTG is calling those Naga instead of the Mowzie Naga like it is supposed to be doing.

**Comment by PG85** [comment]
Hey @Gunner76th , if you're configuring mobs in the biome config via their full resourcelocation name (modname:mobname), then there shouldn't be a way for OTG to fetch the wrong mob (each resourcelocation registered should be unique). Judging by MCPitman's feedback, it sounds more like CustomMobSpawner/BetterAnimals is refering to mobs by a short name, instead of the full name, and so fetching the wrong mob if there are multiple with the same short name?

**Comment by Jishuna** [comment]
> Hey @Gunner76th , if you're configuring mobs in the biome config via their full resourcelocation name (modname:mobname), then there shouldn't be a way for OTG to fetch the wrong mob (each resourcelocation registered should be unique). Judging by MCPitman's feedback, it sounds more like CustomMobSpawner/BetterAnimals is refering to mobs by a short name, instead of the full name, and so fetching the wrong mob if there are multiple with the same short name?

We discussed this a lot in the discord last night, you should take a look at what we were able to find.

**Comment by PG85** [comment]
Ok, lets continue this in the Disc for now.
